### PR TITLE
[java] Java 16 support

### LINKED ISSUE #3079: [java] Support JDK 16
[OpenJDK 16](http://openjdk.java.net/projects/jdk/16/) GA is scheduled for 2021/03/16. So we should implement
support with PMD 6.32.0

The following JEPs are important for us:

*   JEP 394: [Pattern Matching for instanceof](https://openjdk.java.net/jeps/394)
    *   Was in JDK14+JDK15 a preview feature
    *   In JDK16 it will be a standard feature
*   JEP 395: [Records](https://openjdk.java.net/jeps/395)
    *   Was in JDK14+JDK15 a preview feature
    *   In JDK16 it will be a standard feature
*   JEP 397: [Sealed Classes (Second Preview)](https://openjdk.java.net/jeps/397)

In summary: https://javaalmanac.io/jdk/16/

This issue is similar to the previous: #2646

EA Release of JDK16 is available here: http://jdk.java.net/16/

Tasks: see #3111

### LINKED ISSUE #3093: [core] Update beta version of asm with current version
**Is your feature request related to a problem? Please describe.**
In my build beta and alpha version are not allowed to ensure stability.
This is not really a problem but stable versions are better that beta.

**Describe the solution you'd like**
PMD is using a beta version on org.ow2.asm:asm:9.0-beta
Actually version 9 is available https://mvnrepository.com/artifact/org.ow2.asm/asm/9.0
Can you please update asm to the latest stable version?

### PR DESCRIPTION
**Pull request by adangel**
## Describe the PR

Adds Java 16 support

- [x] Add new java language versions 16 and 16-preview. Also add in KotlinTestingDsl.kt.
- [x] Update asm dependency: https://search.maven.org/artifact/org.ow2.asm/asm
- [x] Promote "Pattern Matching for instanceof" (394) to a standard feature. Note, there might be some small changes from last preview.
- [x] Promote "Records" (395) to a standard feature. Note, there might be some small changes from last preview.
- [x] Update "Sealed Classes" (397) preview implementation. Note, there might be some small changes from last preview.
- [x] Remove java language version 14-preview
- [x] Write release notes
    * ASTTypeTestPattern renamed to ASTTypePattern
    * Not Experimental anymore (ASTVariableDeclarationId::isPatternBinding, ASTPattern, ASTTypePattern)
    * Renamed ASTRecordConstructorDeclaration to ASTCompactConstructorDeclaration to align naming to JLS
    * ASTRecordDeclaration, ASTRecordComponentList, ASTRecordComponent, ASTRecordBody, ASTCompactConstructorDeclaration are not longer Experimental
    * ...
- [x] Prepare merge for PMD 7 --> #3120

## Related issues

- Fixes #3079
- Fixes #3093 

## Ready?

- [x] Added unit tests for fixed bug/feature
- [x] Passing all unit tests
- [x] Complete build `./mvnw clean verify` passes (checked automatically by travis)
- [x] Added (in-code) documentation (if needed)

### CODE REVIEW

**Comment by pmd-test** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="true">This changeset changes 0 violations,<br />
introduces 0 new violations, 0 new errors and 0 new configuration errors,<br />
removes 0 violations, 0 errors and 0 configuration errors.<br />
<a href="https://chunk.io/pmd/a67d4c0887714d6ead36045d163efba0/diff/index.html">Full report</a></td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: pmd-java/src/test/java/net/sourceforge/pmd/lang/java/ast/Java16PreviewTreeDumpTest.java [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+    }
+
+    @Test(expected = ParseException.class)
+    public void sealedClassBeforeJava16Preview() {
+        java16.parseResource("geometry/Shape.java");
+    }
+
+    @Test
+    public void sealedClass() {
+        doTest("geometry/Shape");
```
Comment (oowekyala) [inline pmd-java/src/test/java/net/sourceforge/pmd/lang/java/ast/Java16PreviewTreeDumpTest.java:37]: It didn't occur to me that we could use a path, that's nice

File: pmd-java/src/main/java/net/sourceforge/pmd/lang/java/ast/ASTPermitsList.java [hunk 2]
```diff
@@ -12,14 +12,14 @@
...
+ * <p>This is a Java 15 Preview and Java 16 Preview feature.
  *
- * <p>See https://openjdk.java.net/jeps/360
+ * <p>See https://openjdk.java.net/jeps/397
  *
  * <pre class="grammar">
  *
- *  PermittedSubclasses ::= "permits" (TypeAnnotation)* ClassOrInterfaceType
- *                ( "," (TypeAnnotation)* ClassOrInterfaceType )*
+ *  PermittedSubclasses ::= "permits" ClassOrInterfaceType
```
Comment (oowekyala) [inline pmd-java/src/main/java/net/sourceforge/pmd/lang/java/ast/ASTPermitsList.java:21]:
```suggestion
 *  PermitsList ::= "permits" ClassOrInterfaceType
```

**Review by oowekyala — APPROVED** [review]
Thanks, this looks good :)

**Pull request merged**